nni/retiarii/utils.py

```diff
@@ -115,43 +117,78 @@ def top(cls, key: str) -> Any:
 
 class ModelNamespace:
     """
-    To create an individual namespace for models to enable automatic numbering.
+    To create an individual namespace for models:
+
+    1. to enable automatic numbering;
+    2. to trace general information (like creation of hyper-parameters) of model.
+
+    A namespace is bounded to a key. Namespace bounded to different keys are completed isolated.
+    Namespace can have sub-namespaces (with the same key). The numbering will be chained (e.g., ``model_1_4_2``).
     """
 
     def __init__(self, key: str = _DEFAULT_MODEL_NAMESPACE):
         # for example, key: "model_wrapper"
         self.key = key
 
+        # the "path" of current name
+        # By default, it's ``[]``
+        # If a ``@model_wrapper`` is nested inside a model_wrapper, it will become something like ``[1, 3, 2]``.
+        # See ``__enter__``.
+        self.name_path: List[int] = []
+
+        # parameter specs.
+        # Currently only used trace calls of ModelParameterChoice.
+        self.parameter_specs: List[ParameterSpec] = []
+
     def __enter__(self):
         # For example, currently the top of stack is [1, 2, 2], and [1, 2, 2, 3] is used,
         # the next thing up is [1, 2, 2, 4].
         # `reset_uid` to count from zero for "model_wrapper_1_2_2_4"
         try:
-            current_context = ContextStack.top(self.key)
-            next_uid = uid(self._simple_name(self.key, current_context))
-            ContextStack.push(self.key, current_context + [next_uid])
-            reset_uid(self._simple_name(self.key, current_context + [next_uid]))
+            parent_context: 'ModelNamespace' = ModelNamespace.current_context(self.key)
+            next_uid = uid(parent_context._simple_name())
```

Comment:
why using uid rather than string (e.g., module name)

Reply:
We can't get module name because they are assigned in the parent module. At least not without hacks like `inspect.get_frame`.

Instead, uid is simple and globally unique.